src/cpp/core/include/core/FileLock.hpp

```diff
@@ -17,30 +17,90 @@
 #define CORE_FILE_LOCK_HPP
 
 #include <boost/utility.hpp>
+#include <boost/shared_ptr.hpp>
 #include <boost/scoped_ptr.hpp>
+#include <boost/asio.hpp>
+#include <boost/function.hpp>
+#include <boost/bind.hpp>
 
 namespace rstudio {
 namespace core {
 
+namespace locks {
+void initialize();
+} // namespace locks
+
 class Error;
 class FilePath;
 
 class FileLock : boost::noncopyable
 {
 public:
-   static bool isLocked(const FilePath& lockFilePath);
+   
+   enum LockType { LOCKTYPE_ADVISORY, LOCKTYPE_LINKBASED };
+   
+   // control what kind of locks are generated by 'create()' method
+   static void setLockType(LockType type) { type_ = type; }
+   static LockType getLockType() { return type_; }
+   static boost::shared_ptr<FileLock> create();
```

Comment:
I think we need both a `create(LockType)` method and a `createDefault` method so that both types can potentially co-exist in a process